### PR TITLE
[SFSRAP-37] Added jar manifest builder

### PR DESCRIPTION
**Pull request by duncanpMS**
Re-added check that rules are generated as expected

I've used the modified exe to create plugins for three third-party analyzers, which I've tested end-to-end.

### CODE REVIEW

File: PluginGenerator/JarManifestBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+using System.IO;
+using System.Linq;
+using System.Text;
+
+namespace SonarQube.Plugins
+{
+    /// <summary>
+    /// Builds a valid v1.0 Jar Manifest.
+    /// See http://docs.oracle.com/javase/6/docs/technotes/guides/jar/jar.html#JAR%20Manifest
+    /// </summary>
```
Comment (duncanpMS) [inline PluginGenerator/JarManifestBuilder.cs]: Previously, we were calling the Java *jar* utility and it was building valid manifest files for us. Now we're manually creating them, so we're responsible for ensuring they are valid.
The main issue is with splitting long values across multiple lines.

File: RoslynPluginGenerator/AnalyzerPluginGenerator.cs [hunk 2]
```diff
@@ -398,7 +398,7 @@ private string BuildPlugin(RoslynPluginDefinition definition, string outputDirec
                         .SetRepositoryKey(repoKey)
                         .SetRepositoryName(definition.Manifest.Name)
                         .SetRulesFilePath(definition.RulesFilePath)
-                        .SetManifestProperties(definition.Manifest)
+                        .SetPluginManifestProperties(definition.Manifest)
```
Comment (duncanpMS) [inline RoslynPluginGenerator/AnalyzerPluginGenerator.cs]: Renamed to make it clearer that this doesn't refer to the jar manifest.

File: RoslynPluginGenerator/RoslynPluginJarBuilder.cs [hunk 3]
```diff
@@ -53,23 +53,23 @@ public RoslynPluginJarBuilder(ILogger logger)
         {
             if (logger == null)
             {
-                throw new ArgumentNullException("logger");
+                throw new ArgumentNullException(nameof(logger));
```
Comment (duncanpMS) [inline RoslynPluginGenerator/RoslynPluginJarBuilder.cs]: Changed this file to use "nameof(x)"

File: RoslynPluginGenerator/RoslynPluginJarBuilder.cs [hunk 4]
```diff
@@ -329,46 +324,26 @@ private static string GetAbsoluteResourcePath(string relativeFilePath)
...
+        
         /// <summary>
         /// Sets the invariant, required manifest properties
         /// </summary>
-        private void EnsureCoreManifestPropertiesExist()
+        private void SetFixedManifestProperties()
         {
             // This property must appear first in the manifest.
             // See http://docs.oracle.com/javase/6/docs/technotes/guides/jar/jar.html#JAR%20Manifest
-            this.manifestProperties.Add("Manifest-Version", "1.0");
```
Comment (duncanpMS) [inline RoslynPluginGenerator/RoslynPluginJarBuilder.cs]: This is now added by the manifest builder.

File: Tests/IntegrationTests/Roslyn/RoslynGenTests.cs [hunk 5]
```diff
@@ -157,27 +159,27 @@ private static void AssertExpectedPropertyDefinitionValue(string propertyName, s
...
-            //Assert.AreEqual(descriptor.Title.ToString(), actual.Name, "Unexpected rule name");
-            //Assert.AreEqual(descriptor.Id, actual.Key, "Unexpected rule key");
-            //AssertPropertyHasValue(actual.Severity, "Severity");
-        //}
+            Rule actual = matches.Single();
+            Assert.AreEqual(descriptor.Title.ToString(), actual.Name, "Unexpected rule name");
+            Assert.AreEqual(descriptor.Id, actual.Key, "Unexpected rule key");
+
+            Assert.IsNotNull(actual.Severity, "Severity should be specified");
+        }
```
Comment (duncanpMS) [inline Tests/IntegrationTests/Roslyn/RoslynGenTests.cs]: Re-instated the old checks.

File: PluginGenerator/JarManifestBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+        public bool TryGetValue(string name, out string value)
+        {
+            return this.properties.TryGetValue(name, out value);
+        }
+
+        /// <summary>
+        /// Builds and returns the manifest
+        /// </summary>
+        public string GetManifest()
```
Comment (bgavrilMS) [inline PluginGenerator/JarManifestBuilder.cs]: BuildManifest() might be a better name, GetX suggests a simple fetch / query operation

File: PluginGenerator/JarManifestBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        /// <summary>
+        /// Writes an entry, splitting it across multiple lines if required
+        /// </summary>
+        /// <remarks>NOTE: we assuming there are no multi-byte characters in the value. If there are,
+        /// the string won't be split correctly</remarks>
+        private static void WriteEntry(string key, string value, StringBuilder sb)
+        {
+            string text = string.Format(System.Globalization.CultureInfo.InvariantCulture, "{0}: {1}", key, value);
+
+            Debug.Assert(!value.Any(c => char.IsSurrogate(c)), "Not expecting multi-byte characters");
```
Comment (bgavrilMS) [inline PluginGenerator/JarManifestBuilder.cs]: Should this result in a user exception - i.e. only en-us chars are supported? I'm wondering what would happen if someone were to actually encounter this - it's fine to fail, but will we ever understand the root cause?

**Comment by bgavrilMS** [comment]
LGTM

File: PluginGenerator/JarManifestBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+            // Write the first line without a continuation character
+            sb.AppendLine(text.Substring(0, MaxLineLength));
+            text = text.Substring(MaxLineLength);
+
+            // For every remaining line, we need to reserve characters
+            // for the continuation prefix
+            const string continuationPrefix = " ";
+            int maxActualCharsPerLine = MaxLineLength - continuationPrefix.Length;
+            
+            while (text.Length > maxActualCharsPerLine)
```
Comment (dbolkensteyn) [inline PluginGenerator/JarManifestBuilder.cs]: it should be possible to change this condition into `while (!text.IsEmpty()) { .. }` and get rid of the `Write out whatever is left` below

File: PluginGenerator/JarManifestBuilder.cs (hunk 7 above)
Comment (dbolkensteyn) [inline PluginGenerator/JarManifestBuilder.cs]: I also was a bit puzzled, but surrogate seem to affect only very few unicode characters, i.e. the ones that don't fit on a single UTF-16 `char`?

French accents for examples are still going to be supoorted I think

**Comment by dbolkensteyn** [comment]
LGTM

File: PluginGenerator/JarManifestBuilder.cs (hunk 7 above)
Comment (duncanpMS) [inline PluginGenerator/JarManifestBuilder.cs]: UTF-8 is a variable width encoding: most characters fit into a single byte, but not all. We're not limited to _en-us_; it'll work with any single-byte characters.

The assertion is trying to check the assumption stated in the _remarks_ i.e. that there are no multi-byte characters. However, I wasn't certain that the check I was using was correct so I didn't want to make it an exception. I also wanted to see if this occurred in any of the real-world analyzers during testing (it hasn't so far).
It affects few enough characters that I think I'll leave it as it is for the time being.

File: PluginGenerator/JarManifestBuilder.cs (hunk 6 above)
Comment (duncanpMS) [inline PluginGenerator/JarManifestBuilder.cs]: Done

File: PluginGenerator/JarManifestBuilder.cs (hunk 8 above)
Comment (duncanpMS) [inline PluginGenerator/JarManifestBuilder.cs]: Done

**Pull request merged**